parser-typechecker/src/Unison/PatternMatchCoverage.hs

```diff
@@ -0,0 +1,82 @@
+--
+-- 1. [Desugar]("Unison.PatternMatchCoverage.Desugar") a match expression into a 'Unison.PatternMatchCoverage.GrdTree.GrdTree'.
+-- 2. Annotate the @GrdTree@ leaves with [refinement types]("Unison.PatternMatchCoverage.NormalizedConstraints")
+-- describing values that match this branch. Redundant and inaccessible patterns are then identified by @GrdTree@ leaves
+-- with uninhabited refinement types. Inaccessible patterns are distinguished by an effect being performed between the
+-- @GrdTree@ root and the leaf.
+-- 3. Traverse the @GrdTree@ building up a refinement type describing uncovered values. If the resulting refinement type
+-- is inhabited then the match is missing clauses.
+-- 4. Find inhabitants of the uncovered refinement type to present to the user.
+--
+-- Step (1) is implemented by 'desugarMatch'. Steps (2) and (3) are
+-- implemented as a single traversal: 'uncoverAnnotate'/'classify'. Step (4) is
+-- implemented by 'expandSolution'/'generateInhabitants'.
+module Unison.PatternMatchCoverage
+  ( checkMatch,
+  )
+where
+
+import qualified Data.Set as Set
+import Debug.Trace
+import Unison.Pattern (Pattern)
+import Unison.PatternMatchCoverage.Class (Pmc (..))
+import Unison.PatternMatchCoverage.Desugar (desugarMatch)
+import Unison.PatternMatchCoverage.GrdTree (prettyGrdTree)
+import qualified Unison.PatternMatchCoverage.NormalizedConstraints as NC
+import Unison.PatternMatchCoverage.PmGrd (prettyPmGrd)
+import Unison.PatternMatchCoverage.Solve (classify, expandSolution, generateInhabitants, uncoverAnnotate)
+import qualified Unison.Term as Term
+import qualified Unison.Type as Type
+import qualified Unison.Util.Pretty as P
+
+-- | Perform pattern match coverage checking on a match expression
+checkMatch ::
+  forall vt v loc m.
+  (Pmc vt v loc m) =>
+  -- | the match location
+  loc ->
+  -- | scrutinee type
+  Type.Type vt loc ->
+  -- | match cases
+  [Term.MatchCase loc (Term.Term' vt v loc)] ->
+  -- | (redundant locations, inaccessible locations, inhabitants of uncovered refinement type)
+  m ([loc], [loc], [Pattern ()])
+checkMatch matchLocation scrutineeType cases = do
+  v0 <- fresh
+  grdtree0 <- desugarMatch matchLocation scrutineeType v0 cases
+  (uncovered, grdtree1) <- uncoverAnnotate (Set.singleton (NC.declVar v0 scrutineeType id NC.emptyNormalizedConstraints)) grdtree0
+  uncoveredExpanded <- concat . fmap Set.toList <$> traverse (expandSolution v0) (Set.toList uncovered)
+  let sols = map (generateInhabitants v0) uncoveredExpanded
+  let (_accessible, inaccessible, redundant) = classify grdtree1
+  let debugOutput =
+        P.sep
+          "\n"
+          [ P.hang "desugared:" (prettyGrdTree prettyPmGrd (\_ -> "<loc>") grdtree0),
+            P.hang "annotated:" (prettyGrdTree NC.prettyDnf (NC.prettyDnf . fst) grdtree1),
+            P.hang "uncovered:" (NC.prettyDnf uncovered),
+            P.hang "uncovered expanded:" (NC.prettyDnf (Set.fromList uncoveredExpanded))
+          ]
+      shouldDebug = False
```

Comment:
BTW if you'd like to leave the debugging information in, you can add a new case to our [DebugFlags](https://github.com/unisonweb/unison/blob/8eaeeefb1959ed29c53f4ba82c2a436fd473ecbf/lib/unison-prelude/src/Unison/Debug.hs#L25) and use the debugging helpers there.


Then can display debugging information any time with `UNISON_DEBUG=PatternMatchChecker` or w/e

Reply:
(Also it prints using a "pretty" trace which is always nice for complex types).